pkg/common/scheduler/statefulset/scheduler.go

```diff
@@ -160,73 +197,237 @@ func (s *StatefulSetScheduler) Schedule(vpod scheduler.VPod) ([]duckv1alpha1.Pla
+func (s *StatefulSetScheduler) addReplicas(state *state, diff int32, placements []duckv1alpha1.Placement, evenSpread int32) ([]duckv1alpha1.Placement, int32) {
+	// Pod affinity MaxFillup algorithm (evenSpread == 0): prefer adding replicas to existing pods to fill them up before adding to new pods
+	// Pod affinity EvenSpread algorithm (evenSpread > 0): spread replicas across pods in different regions for HA
 	newPlacements := make([]duckv1alpha1.Placement, 0, len(placements))
+	logger := s.logger.With("addReplicas()", zap.Int32("schedule with even spread:", evenSpread))
+
+	if evenSpread > 0 {
+		placementsByZone := getPlacementsByZoneKey(placements)
+		logger.Infow("addReplicas()", zap.Int32("diff:", diff), zap.Any("placementsByZone:", placementsByZone))
+
+		for zoneName, placementOrdinals := range placementsByZone { //iterate through each zone
+			totalInZone := getTotalVReplicasInZone(placements, zoneName)
+			logger.Infow("addReplicas() to existing", zap.String("zoneName", zoneName), zap.Int32("totalInZone:", totalInZone))
+
+			for _, ordinal := range placementOrdinals { //iterating through all existing pod belonging to a single zone
+				placement := s.getPlacementFromPodOrdinal(placements, ordinal)
+
+				// Is there space in Pod?
+				f := state.Free(ordinal)
+				if diff >= 0 && f > 0 && totalInZone < evenSpread {
+					allocation := integer.Int32Min(diff, integer.Int32Min(f, (evenSpread-totalInZone)))
+					logger.Infow("addReplicas() to existing", zap.Int32("allocation", allocation))
+
+					newPlacements = append(newPlacements, duckv1alpha1.Placement{
+						PodName:   placement.PodName,
+						ZoneName:  zoneName,
+						VReplicas: placement.VReplicas + allocation,
+					})
+
+					diff -= allocation
+					state.SetFree(ordinal, f-allocation)
+					totalInZone += allocation
+					logger.Infow("addReplicas() to existing", zap.Int32("diff", diff), zap.Int32("totalInZone", totalInZone))
+				} else {
+					newPlacements = append(newPlacements, placement)
+				}
+			}
+		}
 
-	// Add to existing
-	for i := 0; i < len(placements); i++ {
-		podName := placements[i].PodName
-		ordinal := ordinalFromPodName(podName)
-
-		// Is there space in PodName?
-		f := state.Free(ordinal)
-		if diff >= 0 && f > 0 {
-			allocation := integer.Int32Min(f, diff)
-			newPlacements = append(newPlacements, duckv1alpha1.Placement{
-				PodName:   podName,
-				VReplicas: placements[i].VReplicas + allocation,
-			})
-
-			diff -= allocation
-			state.SetFree(ordinal, f-allocation)
-		} else {
-			newPlacements = append(newPlacements, placements[i])
+		if diff > 0 {
+			for ordinal := int32(0); ordinal < s.replicas; ordinal++ {
+				f := state.Free(ordinal)
+				logger.Infow("Trying to schedule on new pods", zap.Int32("ordinal", ordinal), zap.Int32("free", f))
+				if f > 0 { //here it is possible to hit pods that are in existing placements
+					podName, zoneName, err := s.getZoneNameFromPodOrdinal(ordinal)
```

Comment:
Is it possible to have this info as part of the state? I'm worried about race condition here.

Reply:
also it's not clear what to do when there is an error. It shouldn't be `continue`

Reply:
yeah, I see ```s.podClient.Get(context.Background(), podName, metav1.GetOptions{})``` erroring here and there with this error ```"error":"pods \"kafkasource-mt-adapter-4\" not found",``` and then most times during the next reconcile, it passes. Finding pod fails which means we dont get node info and then the zone info.

Reply:
Storing this in state is a good idea. I will try that too.